### PR TITLE
[KTIJ-31555] [kotlin] Fixed cross-language ChangeAnnotationAction for java annotations used in kotlin code.

### PR DESCRIPTION
**Pull request by Rollczi**
Fixes: [KTIJ-31555](https://youtrack.jetbrains.com/issue/IDEA-360131/kotlinidea-Fixed-cross-language-ChangeAnnotationAction-for-java-annotations-used-in-kotlin-code)

EDIT: It would be nice to get a `hacktoberfest-accepted` label too 😄 

This PR fixes the problem with `JvmElementActionsFactory#createChangeAnnotationAttributeActions`. 
When executing `invoke`, if we modify, for example, a class in Kotlin that uses java annotations, then the attributes are not assigned correctly. I found this bug while migrating to K2.

Java Annotations work differently than Kotlin Annotations in Kotlin code. Because varargs for an array do not exist in Kotlin annotations. After all, kotlin supports java syntax if annotation is from a Java source file.

### IDE Opened project
`Permission.java`
```java
public @interface Permission {
    String[] value();
}
```
`SomeCode.kt`
```kt
@Permission("before1", "before2", "before3")
```

### IDE Plugin
```java
JvmElementActionsFactory factory = // ...;
List<IntentionAction> actions = factory.createChangeAnnotationAttributeActions(
    permissionAnnotation,
    -1,
    new AnnotationAttributeRequest(
        "value",
        new ArrayValue(
            List.of(
                new StringValue("after1"),
                new StringValue("after2"),
                new StringValue("after3")
            )
        )
    ),
    "Change value",
    "Change value"
);

for (IntentionAction intentionAction : actions) {
    intentionAction.invoke(project, new ImaginaryEditor(project, document), annotation.getContainingFile());
}
```

##  Before my changes
`SomeCode.kt`
```kt
@Permission("before1", "before2", "before3", "after1")
```

## After my changes
`SomeCode.kt`
```kt
@Permission("after1", "after2", "after3")
```

### CODE REVIEW

**Comment by akozlova** [comment]
The change is done for the old plugin, K1 mode. It is going to be disabled by default in one of the nearest releases. Would you like to add corresponding code to the K2 mode (`org.jetbrains.kotlin.idea.k2.codeinsight.quickFixes.createFromUsage.K2ElementActionsFactory` entry point)? Thanks

**Comment by Rollczi** [comment]
I can add (copy&paste) corresponding code to the K2 implementation, but I think there should be one implementation that works with both modes (K1 and K2) and this should be reimplemented in other PR.


It's a bit strange that someone adds one of their own use cases and the rest are ignored and not supported e.g.:
image

Also idk how to replace `ShortenReferences.DEFAULT.process(annotationEntry)`

File: plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt [hunk 1]
```diff
@@ -114,9 +118,189 @@ class K2ElementActionsFactory : JvmElementActionsFactory() {
...
+                //ShortenReferences.DEFAULT.process(annotationEntry)
+                return
+            }
+
+            when (language) {
+                JavaLanguage.INSTANCE -> changeJava(annotationEntry, argumentList, dummyArgumentList)
+                KotlinLanguage.INSTANCE -> changeKotlin(annotationEntry, argumentList, dummyArgumentList)
+                else -> changeKotlin(annotationEntry, argumentList, dummyArgumentList)
+            }
+            //ShortenReferences.DEFAULT.process(annotationEntry)
```
Comment (akozlova) [inline plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt:186]: there is `shortenReference` function or `org.jetbrains.kotlin.idea.base.codeInsight.ShortenReferencesFacility` for the common code

File: plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt [hunk 2]
```diff
@@ -114,9 +118,189 @@ class K2ElementActionsFactory : JvmElementActionsFactory() {
...
+    //TODO: this is copy paste from KotlinElementActionsFactory, maybe we should extract it to common place?
+    // or implement this for K1 and K2?
+
+    private class ChangeAnnotationAction(
+        annotationEntry: KtAnnotationEntry,
+        private val attributeIndex: Int,
+        private val request: AnnotationAttributeRequest,
+        @IntentionName private val text: String,
+        @IntentionFamilyName private val familyName: String
+    ) : IntentionAction {
```
Comment (akozlova) [inline plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt:143]: in K2 mode, we try to implement intentions through `org.jetbrains.kotlin.idea.codeinsight.api.applicable.intentions.KotlinPsiUpdateModCommandAction`. You won't need to implement preview and it's easier to reuse the code for different headless applications

File: plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt [hunk 3]
```diff
@@ -114,9 +118,189 @@ class K2ElementActionsFactory : JvmElementActionsFactory() {
...
+            return IntentionPreviewInfo.DIFF
+        }
+
+        override fun invoke(project: Project, editor: Editor?, file: PsiFile) {
+            val annotationEntry = pointer.element ?: return
+            invokeImpl(annotationEntry, project)
+        }
+
+        private fun invokeImpl(annotationEntry: KtAnnotationEntry, project: Project) {
+            val facade = JavaPsiFacade.getInstance(annotationEntry.project)
```
Comment (akozlova) [inline plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt:167]: BTW project is passed as parameter, no need to retrieve it from psi

File: plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt (hunk 2 above)
Comment (Rollczi) [reply plugins/kotlin/code-insight/kotlin.code-insight.k2/src/org/jetbrains/kotlin/idea/k2/codeinsight/quickFixes/createFromUsage/K2ElementActionsFactory.kt:143]: Okay but I can't extract this class to `kotlin.code-insight.intentions.k2`

**Comment by Rollczi** [comment]
Hi @akozlova

Will be possible to add `hacktoberfest-accepted` label ? 😄

**Pull request closed without merging**